File: ImportRegistry.py
# -*- coding: utf-8 -*-
"""
Created on Fri Mar 18 09:12:08 2022

@author: naqavi
"""

import pandas as pd
import numpy as np

df = pd.read_csv("personbilaritrafik.csv", encoding = 'latin1')

# produce long file with total fleet
cols = ['Bensin','Diesel','El','Etanol.hybrid','Bensin.El','Gas','Övriga','Totalt','Miljöbil']
x = df['Bensin'].append(df['Diesel']).append(df['El']).append(
    df['Etanol.hybrid']).append(df['Bensin.El']).append(df['Gas']).append(
    df['Övriga']).append(df['Totalt']).append(df['Miljöbil']).reset_index(drop = True)
new_cols = ['Year', 'Komkod','Kommun','ee.index','rr.index']
new = df.filter(new_cols, axis=1)
Registry_tot = pd.DataFrame(data = np.tile(new,(9,1)), columns = new_cols) 
Registry_tot['Drivmedel'] = np.repeat(cols,182)
Registry_tot['tot_veh'] = x
del x, new_cols

# produce long file with fleet subset from 2009 onward
x = df['Bensin.from09'].append(df['Diesel.from09']).append(df['El.from09']).append(
    df['Etanol.hybrid.from09']).append(df['Bensin.El.from09']).append(df['Gas.from09']).append(
    df['Övriga.from09']).append(df['Totalt.from09']).append(df['Miljöbil.from09']).reset_index(drop = True)
new_cols = ['Year', 'Komkod','Kommun']
new = df.filter(new_cols, axis=1)
Registry_f09 = pd.DataFrame(data = np.tile(new,(9,1)), columns = new_cols) 
Registry_f09['Drivmedel'] = np.repeat(cols,182)
Registry_f09['f09_veh'] = x
del x, new_cols, new

# merge
registry = pd.merge(Registry_tot, Registry_f09, how='left', on=["Year","Komkod","Kommun","Drivmedel"])
registry['f09_pv'] = registry['f09_veh']/registry['tot_veh']
registry['t08_pv'] = (registry['tot_veh']-registry['f09_veh'])/registry['tot_veh']

# Modify Drivmedel in "registry" to conform to "passages"
registry['Drivmedel'] = registry['Drivmedel'].str.replace('Bensin', 'Petrol', regex = True)
registry['Drivmedel'] = registry['Drivmedel'].str.replace('Diesel', 'Diesel', regex = True)
registry['Drivmedel'] = registry['Drivmedel'].str.replace('El', 'Electric', regex = True)
registry['Drivmedel'] = registry['Drivmedel'].str.replace('Bensin.El', 'Electric.Hybrid', regex = True)
registry['Drivmedel'] = registry['Drivmedel'].str.replace('Etanol.hybrid', 'Ethanol', regex = True)
registry['Drivmedel'] = registry['Drivmedel'].str.replace('Bensin', 'Petrol', regex = True)
registry['Drivmedel'] = registry['Drivmedel'].str.replace('Gas', 'Natural.Gas', regex = True)
registry['Drivmedel'] = registry['Drivmedel'].str.replace('Övriga', 'Other', regex = True)
registry['Drivmedel'] = registry['Drivmedel'].str.replace('Totalt', 'Total', regex = True)
registry['Drivmedel'] = registry['Drivmedel'].str.replace('Miljöbil', 'AFV', regex = True)

# Get share of clean vehicles pre-2008 for all-county
x = registry['t08_pv'].loc[registry['Drivmedel'] == 'AFV']
w = registry['tot_veh'].loc[registry['Drivmedel'] == 'AFV']
share_clean_t08 = sum(x*w)/(sum(w))

x = registry['f09_pv'].loc[registry['Drivmedel'] == 'AFV']
w = registry['tot_veh'].loc[registry['Drivmedel'] == 'AFV']
share_clean_f09 = sum(x*w)/(sum(w))

File: ImportRegistry.R

# Import vehicle registry data for 2012

registry.wide <- read.csv("personbilaritrafik.csv", header=TRUE)

# produce long file with total fleet
registry.tot <- registry.wide %>%
  dplyr::select(Year, Komkod, Kommun, Bensin, Diesel, El, 
                Etanol.hybrid, Bensin.El, Gas, Övriga, Totalt, Miljöbil, ee.index, rr.index) %>%
  gather(c("Bensin", "Diesel", "El", "Etanol.hybrid", 
           "Bensin.El","Gas","Övriga","Totalt", "Miljöbil"), 
         key="Drivmedel", value="tot.veh", na.rm=TRUE)
# produce long file with fleet subset from 2009 onward
registry.f09 <- registry.wide %>%
  dplyr::select(Year, Komkod, Kommun, Bensin.from09, Diesel.from09, El.from09, 
                Etanol.hybrid.from09, Bensin.El.from09, Gas.from09, Övriga.from09, Totalt.from09, Miljöbil.from09) %>%
  gather(c("Bensin.from09", "Diesel.from09", "El.from09", "Etanol.hybrid.from09", 
           "Bensin.El.from09","Gas.from09","Övriga.from09","Totalt.from09", "Miljöbil.from09"), 
         key="Drivmedel", value="f09.veh", na.rm=TRUE)
# rename fuel types
registry.f09$Drivmedel <- sub(pattern=".from09", replacement="",
                              x=registry.f09$Drivmedel)
# merge
registry <- registry.tot %>%
  left_join(registry.f09, by=c("Year","Komkod","Kommun","Drivmedel"))
#registry$f09.veh[is.na(registry$f09.veh)] <- 0
registry$f09.pv <- registry$f09.veh/registry$tot.veh
registry$t08.pv <- (registry$tot.veh-registry$f09.veh)/registry$tot.veh

#Modify Drivmedel in "registry" to conform to "passages"
registry$Drivmedel <- as.factor(registry$Drivmedel)
registry$Drivmedel <- fct_collapse(registry$Drivmedel,
                                   Petrol = c("Bensin"),
                                   Diesel = c("Diesel"),
                                   Electric = c("El"),
                                   Electric.Hybrid = c("Bensin.El"),
                                   Ethanol = c("Etanol.hybrid"),
                                   Natural.Gas = c("Gas"),
                                   Other = c("Övriga"),
                                   Total = c("Totalt"),
                                   AFV = c("Miljöbil"))

# Get share of clean vehicles pre-2008 for all-county
share.clean.t08 <- weighted.mean(x=registry$t08.pv[registry$Drivmedel=="AFV"], 
                                 w=registry$tot.veh[registry$Drivmedel=="AFV"])
share.clean.f09 <- weighted.mean(x=registry$f09.pv[registry$Drivmedel=="AFV"], 
                                 w=registry$tot.veh[registry$Drivmedel=="AFV"])


